### PR TITLE
Ticket4166 scriptgen rows

### PR DESCRIPTION
**Pull request by Alistair-McGann-Tessella**
### Description of work

Adds data table for script generator actions. Added GUI elements to move/add/duplicate rows.

### Ticket

https://github.com/ISISComputingGroup/IBEX/issues/4166

### Acceptance criteria

- [x] Script generator table can be dropped and re-built for a different action
- [x] Cell values can be manipulated
- [x] Rows can be added, moved, duplicated and deleted
- [x] Changes adequately documented

### Unit tests

Adding, moving and duplicating rows is tested, as is modifying cells in row.
Since the GUI elements added here are tightly bound to these actions, there are no unit tests for 

### Documentation

See updated [wiki page](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/The-IBEX-Script-Generator#data-structures-in-the-gui)

---

#### Code Review

- [x] Is the code of an acceptable quality?
- [x] Do the changes function as described and is it robust?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/JournalViewerView.java [hunk 1]
```diff
@@ -245,7 +247,7 @@ private void changeTableColumns() {
         for (final JournalField field : JournalField.values()) {
             if (model.getFieldSelected(field)) {
                 TableViewerColumn col = table.createColumn(field.getFriendlyName(), 1, true,
-                        new DataboundCellLabelProvider<JournalRow>(table.observeProperty("row")) {
+                        new DataboundCellLabelProvider<JournalRow>(new ObservableMap(Collections.emptyMap())) {
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/JournalViewerView.java]: What was the purpose of this? It seems unrelated to the ticket?

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+	private void addColumn(ActionParameter actionParameter) {
+		String columnName = actionParameter.getName();
+		TableViewerColumn column = createColumn(
+				columnName, 
+				2,
+				new DataboundCellLabelProvider<ScriptGeneratorAction>(observeProperty(columnName)) {
+					@Override
+					protected String stringFromRow(ScriptGeneratorAction row) {
+						// TODO Auto-generated method stub
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java]: Remove auto generated code

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	protected SelectionAdapter getColumnSelectionAdapter(final TableColumn column, final int index) {
+        SelectionAdapter selectionAdapter = new SelectionAdapter() {
+            @Override
+            public void widgetSelected(SelectionEvent e) {
+            }
+        };
+        return selectionAdapter;
+    }
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java]: Documentation on why we're doing this would be good.

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+//		Button button = new Button(grpTable, SWT.NONE);
+//		button.setText("Press me!");
+		
+		//Button btnRowUp = createMoveRowButton(btnsComposite, "move_up.png", "up");
+		
+
+		
+
+
+
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java]: Remove commented code

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        bind();
+		
+        scriptGeneratorModel.addEmptyAction();
+	}
+
+	private void bind() {
+		bindingContext = new DataBindingContext();
+//		bindingContext.bindValue(WidgetProperties.text().observe(lblOrder), 
+//				BeanProperties.value("iteratedNumber").observe(toyModel));
+
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java]: Remove commented code

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+		GridData gdBtnMoveRow = new GridData(SWT.LEFT, SWT.BOTTOM, false, false, 1, 1);
+		gdBtnMoveRow.widthHint = 25;
+		moveButton.setLayoutData(gdBtnMoveRow);
+		moveButton.setImage(ResourceManager.getPluginImage("uk.ac.stfc.isis.ibex.ui", "icons/" + icon));
+        moveButton.setToolTipText("Move selected row " + direction);
+        return moveButton;
+	}
+	
+	/**
+	 * A basic framework to hold the toy interface.
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java]: Comment incorrect

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+		GridData gdQueueContainer = new GridData(SWT.FILL, SWT.FILL, true, false, 1, 1);
+		gdQueueContainer.heightHint = 300;
+		parent.setLayoutData(gdQueueContainer);
+		parent.setLayout(new GridLayout(2, false));
+       
+		table = new ActionsViewTable(parent, SWT.NONE, SWT.SINGLE | SWT.V_SCROLL | SWT.FULL_SELECTION, scriptGeneratorTable);
+		table.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true, 1, 1));
+		
+		//table.addSelectionChangedListener(e -> queueScriptViewModel.setSelectedScript(queuedScriptsViewer.firstSelectedRow()));
+		//addDragAndDrop(queuedScriptsViewer.viewer());
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java]: Is drag and drop in the scope of this ticket? If so do it, if not remove this

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/table/ScriptGeneratorAction.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
+package uk.ac.stfc.isis.ibex.scriptgenerator.table;
+
+import java.util.HashMap;
+import java.util.Map;
+
+
+import uk.ac.stfc.isis.ibex.model.ModelObject;
+
+/**
+ * Class defines one action or 'step' in the script
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/table/ScriptGeneratorAction.java]: End with a full stop otherwise checkstyle complains, stupid I know. You should double check the other docstrings you've added too

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java (hunk 3 above)
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java]: This method no longer useful, it was a first attempt at stopping the columns from being reordered. This is done by overriding `ColumnComparator`.

**Comment by Alistair-McGann-Tessella** [comment]
> 
> 
> I'm also not sure of the purpose behind the refactor of `BooleanWritableObserableAdapter` and `uk.ac.stfc.isis.ibex.ui.widgets.epics`, were they required as part of this ticket?

We made these changes to make sure that our dependencies wouldn't snowball when we imported some classes. I think `uk.ac.stfc.isis.ibex.ui.widgets.StringEditingSupport` would otherwise have caused us to import everything from epics

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        final Button btnInsertAction = new Button(actionsControlsGrp, SWT.NONE);
+        btnInsertAction.setText("Add Action");
+        btnInsertAction.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true, 1, 1));
+
+		btnInsertAction.addSelectionListener(new SelectionAdapter() {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				scriptGeneratorModel.addEmptyAction();
+			}
+		});
```
[the remark this answers is not in the document]
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java]: Alistair has swapped this out for all instances in this class

File: base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/JournalViewerView.java (hunk 1 above)
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/JournalViewerView.java]: The JournalViewer table uses an extension of the DataboundTable class. The changes to the underlying table have cause more changes in the JournalViewer in subsequent updates to this branch, so this is no longer an issue.

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java (hunk 2 above)
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java]: Alistair has removed it

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java (hunk 3 above)
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java]: Alistair has removed it

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java (hunk 6 above)
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java]: Alistair has corrected it.

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java (hunk 7 above)
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java]: Alistair has removed it

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java (hunk 4 above)
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java]: Alistair has removed it

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java (hunk 5 above)
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java]: Alistair has removed it

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/table/ScriptGeneratorAction.java (hunk 8 above)
Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/table/ScriptGeneratorAction.java]: Full stop has been added

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/table/ScriptGeneratorAction.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+import uk.ac.stfc.isis.ibex.model.ModelObject;
+
+/**
+ * Class defines one action or 'step' in the script.
+ *
+ */
+public class ScriptGeneratorAction extends ModelObject {
+
+	private HashMap<String, String > actionParameterValues;
```
[the remark this answers is not in the document]
Comment (Alistair-McGann-Tessella) [reply base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/table/ScriptGeneratorAction.java]: I've changed this to ActionParameter and updated the diagram on the [wiki](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/The-IBEX-Script-Generator#data-structures-in-the-gui).

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+    
+    /**
+     * Adds a parameter to this actions table.
+     */
+    @Override
+    protected void addColumns() {    	
+        for (ActionParameter actionParameter: actionsTable.getActionParameters()) {
+			String columnName = actionParameter.getName();
```
Comment (JamesKingWork) [inline base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java]: We should now be using the actionParameter itself

Comment (JamesKingWork) [reply base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java]: Committed change myself as minor.
